Implement VLAN based VF traffic control mechanism

pperiyasamy opened the pull request:
This configures associated VFs for a given unauthenticated PF with reserved VLAN value 4095 and set tx queue to disable state. As soon as first wpa client gets authenticated, it restores original VLAN value and make tx queue state to auto. At the same time it also keeps track of VLAN update on the interface and configure the VF according to PF's authenticated state.

pperiyasamy committed:
Implement VLAN based VF traffic control mechanism

This configures associated VFs for a given unauthenticated PF with reserved VLAN
value 4095 and set tx queue to disable state. As soon as first wpa client gets
authenticated, it restores original VLAN value and make tx queue state to auto.
At the same time it also keeps track of VLAN update on the interface and
configure the VF according to PF's authenticated state.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

lack commented on internal/trafficcontrol/tc.go:
```diff
@@ -189,9 +199,14 @@ func GetSriovVFs(ifName string) ([]string, error) {
 			return nil, fmt.Errorf("failed to get netlink %s: %q", vfIfName, err)
 		}
-		vfNames = append(vfNames, vfLink.Attrs().Name)
+		vfInfo := &VFInfo{Name: vfLink.Attrs().Name, Index: vfLink.Attrs().Index,
```
I think I see something in the code above that's concerning. Basically looking at the directory structure beneath `/sys/class/net/ifname/device/virtfn*`, you *are* guaranteed to have one `virtfn$i` for each [0 <= i <= (sriov_numvfs - 1)], so just iterating over that filepath.Glob result is good. However:
- If the VF is configure as a vfio_pci device, there will be no `net` directory at all, and no VF interface name, and that's okay. We can still run the traffic control like setting the vlan and tx queue state via the `ip` command with only the PF iface name and VF index number.
- If the VF is configured as a netdev device, there is a `net` directory, but if that netdev has been moved into another network namespace (for example, if it's been assigned to another pod already), that net directory will be empty and we won't know the name of the device unless we do some clever searching.  Again, thankfully the new VLAN-based way of controlling traffic doesn't care if we have the VF name, just the base PF interface name and the VF index.

pperiyasamy replied:
Ok, I understand your concern, yes. this won't work considering dpdk and vf is on another net namespace. Now changed it use PF link itself to retrieve VFs. Hope this would work fine for these cases as well. I'm yet to test it to confirm the behavior.

lack replied:
I've done testing on a different SNO (sno-4 in this lab) that checks exactly that - For VFs in other namespaces, and DPDK VFs, does the "set vlan and state" mechanism work, but it would be great to see this in an end-to-end test soon.  I can help set up the containers and SRIOV objects to make it happen if you like.

pperiyasamy replied:
sure, I've tested latest code with VFs (netdevs) and exists on host net namespace, all the basic unauthenticated, authenticated states and vlan change events and stop monitor cases worked fine. When you have few of `ens3f0` VFs into other net namespaces and DPDK mode, then i can verify this one too.

pperiyasamy replied:
@lack I was able to test VFs on other net namespaces and dpdk mode on sno-5 node itself, here is a commit https://github.com/openshift-kni/eapol-operator/pull/12/commits/3a408484d7a6c0181e827128177b40b214a7ef11 needed to make it work.

lack commented on internal/trafficcontrol/tc.go:
```diff
@@ -35,6 +35,14 @@ var (
+	PfName        string
+	Index         int
+	VlanID        int
+	Authenticated bool
```
This `Authenticated`  state really is the property of the PF and the VFs cannot ever differ from each other or from the associated PF.

Instead of tracking it here, is it better to have a link to a parent structure that represents the PF?

pperiyasamy replied:
sure, changed the PF/VF struct to depict authenticated state on the PF.

lack commented on internal/trafficcontrol/tc.go:
```diff
@@ -189,9 +199,14 @@ func GetSriovVFs(ifName string) ([]string, error) {
+		vfInfo := &VFInfo{Name: vfLink.Attrs().Name, Index: vfLink.Attrs().Index,
+			PfName: ifName, Authenticated: false}
+		if len(vfLink.Attrs().Vfs) > 0 {
+			vfInfo.VlanID = vfLink.Attrs().Vfs[0].Vlan
```
We may need to consider how we recover from the case where we set the VLAN to the port to 4095, then we crash, and we recover - How do we know what the "good" VLAN is in that case? May need a separate story...

pperiyasamy replied:
ok, sure.

lack commented on pkg/hostap/monitor.go:
```diff
@@ -240,34 +264,61 @@ func (m *InterfaceMonitor) handleIfEvents() {
+			if linkUpdateEvent.Link.Attrs().Name == m.IfName {
+				m.handlePfEvent(linkUpdateEvent)
+			} else {
+				m.handleVfEvent(linkUpdateEvent)
```
++

I love how clean the code gets when we pull more complicated if clauses out into separate functions!

lack commented on internal/trafficcontrol/tc.go:
```diff
@@ -201,3 +212,33 @@ func IsSriovPF(ifName string) bool {
+	} else {
+		vlan = vfInfo.Vlan
+	}
+	return netlink.LinkSetVfVlan(pfLink, vfInfo.Index, vlan)
```
I think we now need to also call `netlink.LinkSetVfState()` to either `VF_LINK_STATE_AUTO` or `VF_LINK_STATE_DISABLE` depending on the auth state?

pperiyasamy replied:
oh! yes, forgot about link state. changed it now.

lack commented on pkg/hostap/monitor.go:
```diff
@@ -240,36 +261,56 @@ func (m *InterfaceMonitor) handleIfEvents() {
+	defer m.addrMutex.Unlock()
+	for _, vf := range linkUpdateEvent.Link.Attrs().Vfs {
+		if vfInfo, ok := m.PfInfo.VFs[vf.ID]; ok && vfInfo.Vlan != vf.Vlan {
+			vfInfo.Vlan = vf.Vlan
```
only if `vlan != 4095`?

pperiyasamy replied:
good catch ! yes, it avoids an unnecessary netlink call.

lack commented on internal/trafficcontrol/tc.go:
```diff
@@ -201,3 +212,33 @@ func IsSriovPF(ifName string) bool {
+
+func GetSriovPFInfo(ifName string) (*PFInfo, error) {
+	if !IsSriovPF(ifName) {
+		return nil, fmt.Errorf("interface %s is not a sriov pf type", ifName)
```
Maybe we don't need to special-case this any more - Treat all PFs the same, and always populate a `PFInfo` - Just if a PF isn't set up with SRIOV, the VFInfo list will be empty.

pperiyasamy replied:
yes, removed it.

lack commented on pkg/hostap/monitor.go:
```diff
@@ -98,7 +99,18 @@ func (m *InterfaceMonitor) StartMonitor() error {
+		return err
+	}
+	for _, vfInfo := range pfInfo.VFs {
+		err = trafficcontrol.ConfigureVF(m.Logger, pfInfo, vfInfo)
```
I see this model around a lot - monitor.go looping over the VFs.

Would it be cleaner to just pass around the PFInfo struct and have the trafficcontrol code do the looping?

Do we want to make the PFInfo more of an object-oriented object that owns its own actions?

pperiyasamy replied:
yes, correct. now tied appropriate methods to PFInfo/VFInfo structs and using it.

pperiyasamy committed:
Handle dpdk and non root network namespace VFs

This would skip programming tc rules on the DPDK devices and VFs which
are already moved into another network namespace. This is anyway handled
as part of the VF filtering.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

lack commented on internal/trafficcontrol/tc.go:
```diff
@@ -18,15 +18,15 @@ package trafficcontrol
+)
+
+const (
+	ReservedVlan = 4095
```
This probably belongs in the sriov.go file now?

lack commented on pkg/hostap/monitor.go:
```diff
@@ -240,36 +257,58 @@ func (m *InterfaceMonitor) handleIfEvents() {
+		return
+	}
+	for _, vf := range pfLink.Attrs().Vfs {
+		if vfInfo, ok := m.PfInfo.VFs[vf.ID]; ok && vf.Vlan != trafficcontrol.ReservedVlan &&
```
This kind of specialized knowledge feels like it belongs in sriov.go instead of here?

lack commented on pkg/hostap/monitor.go:
```diff
@@ -313,6 +352,13 @@ func (m *InterfaceMonitor) writeCommand(command string) error {
 func (m *InterfaceMonitor) handleAuthenticateEvent(addr string) error {
 	m.addrMutex.Lock()
 	defer m.addrMutex.Unlock()
+	if len(m.authenticatedAddrs) == 0 {
+		m.PfInfo.Authenticated = true
+		err := m.PfInfo.ConfigureVlanStateForVFs()
+		if err != nil {
+			return err
+		}
+	}
```
I would vote for moving this whole block into          `trafficcontrol.AllowTrafficFromMac(m.IfName, addr)`

pperiyasamy replied:
yes, it makes sense. moved this logic into `trafficcontrol.AllowTrafficFromMac `method.

lack commented on pkg/hostap/monitor.go:
```diff
@@ -313,6 +352,13 @@ func (m *InterfaceMonitor) writeCommand(command string) error {
 	defer m.addrMutex.Unlock()
+	if len(m.authenticatedAddrs) == 0 {
+		m.PfInfo.Authenticated = true
+		err := m.PfInfo.ConfigureVlanStateForVFs()
```
Don't we need an opposite of this in `handleDeauthenticateEvent`?

pperiyasamy replied:
yes, done.

lack reviewed (changes requested):
I like this refactoring into `sriov.go` and think it should go further.

I'm okay with waiting on the refactoring until a later commit, but I think the lack of a call to adjust VF vlans on deauthenticate is something we need to fix in this PR
